File: sdk/go/ocsf/schema.go
// Package ocsf provides AITF OCSF Category 7 AI event schema for Go.
//
// OCSF v1.1.0 base objects and AI-specific extension models.
// Based on the OCSF schema from the AITelemetry project, enhanced
// for AITF Category 7 AI events.
package ocsf

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"time"
)

// version is the AITF SDK version used in OCSF metadata.
const version = "1.0.0"

// --- OCSF Enumerations ---

// OCSFSeverity represents OCSF event severity levels.
const (
	SeverityUnknown       = 0
	SeverityInformational = 1
	SeverityLow           = 2
	SeverityMedium        = 3
	SeverityHigh          = 4
	SeverityCritical      = 5
	SeverityFatal         = 6
)

// OCSFStatus represents OCSF event status codes.
const (
	StatusUnknown = 0
	StatusSuccess = 1
	StatusFailure = 2
	StatusOther   = 99
)

// OCSFActivity represents OCSF event activity codes.
const (
	ActivityUnknown = 0
	ActivityCreate  = 1
	ActivityRead    = 2
	ActivityUpdate  = 3
	ActivityDelete  = 4
	ActivityOther   = 99
)

// AIClassUID represents AITF OCSF Category 7 class UIDs.
const (
	ClassUIDModelInference  = 7001
	ClassUIDAgentActivity   = 7002
	ClassUIDToolExecution   = 7003
	ClassUIDDataRetrieval   = 7004
	ClassUIDSecurityFinding = 7005
	ClassUIDSupplyChain     = 7006
	ClassUIDGovernance      = 7007
	ClassUIDIdentity        = 7008
)

// --- OCSF Base Objects ---

// OCSFMetadata holds OCSF event metadata.
type OCSFMetadata struct {
	Version        string            `json:"version"`
	Product        map[string]string `json:"product"`
	UID            string            `json:"uid"`
	CorrelationUID string            `json:"correlation_uid,omitempty"`
	OriginalTime   string            `json:"original_time,omitempty"`
	LoggedTime     string            `json:"logged_time"`
}

// NewOCSFMetadata creates metadata with default values.
func NewOCSFMetadata() OCSFMetadata {
	return OCSFMetadata{
		Version: "1.1.0",
		Product: map[string]string{
			"name":        "AITF",
			"vendor_name": "AITF",
			"version":     version,
		},
		UID:        generateUID(),
		LoggedTime: time.Now().UTC().Format(time.RFC3339),
	}
}

// OCSFActor holds OCSF actor information.
type OCSFActor struct {
	User    map[string]interface{} `json:"user,omitempty"`
	Session map[string]interface{} `json:"session,omitempty"`
	AppName string                 `json:"app_name,omitempty"`
}

// OCSFDevice holds OCSF device/host information.
type OCSFDevice struct {
	Hostname  string            `json:"hostname,omitempty"`
	IP        string            `json:"ip,omitempty"`
	Type      string            `json:"type,omitempty"`
	OS        map[string]string `json:"os,omitempty"`
	Cloud     map[string]string `json:"cloud,omitempty"`
	Container map[string]string `json:"container,omitempty"`
}

// OCSFEnrichment holds OCSF enrichment data.
type OCSFEnrichment struct {
	Name     string `json:"name"`
	Value    string `json:"value"`
	Type     string `json:"type,omitempty"`
	Provider string `json:"provider,omitempty"`
}

// OCSFObservable holds OCSF observable values.
type OCSFObservable struct {
	Name  string `json:"name"`
	Type  string `json:"type"`
	Value string `json:"value"`
}

// --- AI-Specific Extension Models ---

// AIModelInfo holds AI model information.
type AIModelInfo struct {
	ModelID    string                 `json:"model_id"`
	Name       string                 `json:"name,omitempty"`
	Version    string                 `json:"version,omitempty"`
	Provider   string                 `json:"provider,omitempty"`
	Type       string                 `json:"type,omitempty"`
	Parameters map[string]interface{} `json:"parameters,omitempty"`
}

// AITokenUsage holds AI token usage statistics.
type AITokenUsage struct {
	InputTokens     int      `json:"input_tokens"`
	OutputTokens    int      `json:"output_tokens"`
	TotalTokens     int      `json:"total_tokens"`
	CachedTokens    int      `json:"cached_tokens,omitempty"`
	ReasoningTokens int      `json:"reasoning_tokens,omitempty"`
	EstimatedCostUSD *float64 `json:"estimated_cost_usd,omitempty"`
}

// ComputeTotal sets TotalTokens to InputTokens + OutputTokens if it is zero.
func (t *AITokenUsage) ComputeTotal() {
	if t.TotalTokens == 0 {
		t.TotalTokens = t.InputTokens + t.OutputTokens
	}
}

// AILatencyMetrics holds AI operation latency metrics.
type AILatencyMetrics struct {
	TotalMs            float64  `json:"total_ms"`
	TimeToFirstTokenMs *float64 `json:"time_to_first_token_ms,omitempty"`
	TokensPerSecond    *float64 `json:"tokens_per_second,omitempty"`
	QueueTimeMs        *float64 `json:"queue_time_ms,omitempty"`
	InferenceTimeMs    *float64 `json:"inference_time_ms,omitempty"`
}

// AICostInfo holds AI operation cost information.
type AICostInfo struct {
	InputCostUSD  float64 `json:"input_cost_usd"`
	OutputCostUSD float64 `json:"output_cost_usd"`
	TotalCostUSD  float64 `json:"total_cost_usd"`
	Currency      string  `json:"currency"`
}

// AITeamInfo holds multi-agent team information.
type AITeamInfo struct {
	TeamName    string   `json:"team_name"`
	TeamID      string   `json:"team_id,omitempty"`
	Topology    string   `json:"topology,omitempty"`
	Members     []string `json:"members,omitempty"`
	Coordinator string   `json:"coordinator,omitempty"`
}

// AISecurityFinding holds security finding details.
type AISecurityFinding struct {
	FindingType     string   `json:"finding_type"`
	OWASPCategory   string   `json:"owasp_category,omitempty"`
	RiskLevel       string   `json:"risk_level"`
	RiskScore       float64  `json:"risk_score"`
	Confidence      float64  `json:"confidence"`
	DetectionMethod string   `json:"detection_method"`
	Blocked         bool     `json:"blocked"`
	Details         string   `json:"details,omitempty"`
	PIITypes        []string `json:"pii_types,omitempty"`
	MatchedPatterns []string `json:"matched_patterns,omitempty"`
	Remediation     string   `json:"remediation,omitempty"`
}

// ComplianceMetadata holds compliance framework mappings.
type ComplianceMetadata struct {
	NISTAIMRMF map[string]interface{} `json:"nist_ai_rmf,omitempty"`
	MITREAtlas map[string]interface{} `json:"mitre_atlas,omitempty"`
	ISO42001   map[string]interface{} `json:"iso_42001,omitempty"`
	EUAIAct    map[string]interface{} `json:"eu_ai_act,omitempty"`
	SOC2       map[string]interface{} `json:"soc2,omitempty"`
	GDPR       map[string]interface{} `json:"gdpr,omitempty"`
	CCPA       map[string]interface{} `json:"ccpa,omitempty"`
}

// --- OCSF Base Event ---

// AIBaseEvent is the base OCSF event for all AITF Category 7 events.
type AIBaseEvent struct {
	ActivityID  int                `json:"activity_id"`
	CategoryUID int               `json:"category_uid"`
	ClassUID    int                `json:"class_uid"`
	TypeUID     int                `json:"type_uid"`
	Time        string             `json:"time"`
	SeverityID  int                `json:"severity_id"`
	StatusID    int                `json:"status_id"`
	Message     string             `json:"message"`
	Metadata    OCSFMetadata       `json:"metadata"`
	Actor       *OCSFActor         `json:"actor,omitempty"`
	Device      *OCSFDevice        `json:"device,omitempty"`
	Compliance  *ComplianceMetadata `json:"compliance,omitempty"`
	Observables []OCSFObservable   `json:"observables,omitempty"`
	Enrichments []OCSFEnrichment   `json:"enrichments,omitempty"`
}

// ComputeTypeUID computes the type_uid as class_uid * 100 + activity_id.
func (e *AIBaseEvent) ComputeTypeUID() int {
	return e.ClassUID*100 + e.ActivityID
}

// NewAIBaseEvent creates a base event with default values.
func NewAIBaseEvent(classUID, activityID int) AIBaseEvent {
	e := AIBaseEvent{
		ActivityID:  activityID,
		CategoryUID: 7, // AI System Activity
		ClassUID:    classUID,
		Time:        time.Now().UTC().Format(time.RFC3339),
		SeverityID:  SeverityInformational,
		StatusID:    StatusSuccess,
		Metadata:    NewOCSFMetadata(),
	}
	e.TypeUID = e.ComputeTypeUID()
	return e
}

// ToJSON serializes the event to JSON bytes.
func (e *AIBaseEvent) ToJSON() ([]byte, error) {
	return json.Marshal(e)
}

// generateUID creates a cryptographically random unique identifier.
// Uses 16 bytes of randomness (128-bit) for unpredictable UIDs.
func generateUID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		// Fallback to timestamp-based UID if crypto/rand fails
		return time.Now().UTC().Format("20060102150405.000000000")
	}
	return hex.EncodeToString(b)
}

File: sdk/typescript/src/ocsf/schema.ts
/**
 * AITF OCSF Base Schema.
 *
 * OCSF v1.1.0 base objects and AI-specific extension models.
 * Based on the OCSF schema from the AITelemetry project, enhanced
 * for AITF AI events that reuse existing OCSF classes (OCSF PR #1641 /
 * issue #1640).
 */

import { randomUUID } from "crypto";

// --- OCSF Enumerations ---

export enum OCSFSeverity {
  UNKNOWN = 0,
  INFORMATIONAL = 1,
  LOW = 2,
  MEDIUM = 3,
  HIGH = 4,
  CRITICAL = 5,
  FATAL = 6,
}

export enum OCSFStatus {
  UNKNOWN = 0,
  SUCCESS = 1,
  FAILURE = 2,
  OTHER = 99,
}

export enum OCSFActivity {
  UNKNOWN = 0,
  CREATE = 1,
  READ = 2,
  UPDATE = 3,
  DELETE = 4,
  OTHER = 99,
}

/**
 * OCSF category UIDs that AITF AI events map onto.
 *
 * Following OCSF's "reuse existing objects and profiles" approach
 * (OCSF PR #1641 / issue #1640), AITF emits AI telemetry under existing
 * OCSF categories enriched with the `ai_operation` profile, and uses the
 * proposed `ai` category (uid 9) only for genuinely new control-plane
 * classes (agent / delegation lifecycle).
 */
export enum OCSFCategoryUID {
  FINDINGS = 2,
  IAM = 3,
  DISCOVERY = 5,
  APPLICATION = 6,
  AI = 9, // proposed "AI Activity" category (OCSF issue #1640)
}

/**
 * OCSF event class UIDs that AITF AI events map onto.
 *
 * Data-plane AI activity reuses existing OCSF classes; only the agent and
 * delegation control-plane lifecycle use the proposed `ai` category.
 */
export enum OCSFClassUID {
  // Reused existing OCSF classes (verified against the OCSF schema).
  VULNERABILITY_FINDING = 2002,
  COMPLIANCE_FINDING = 2003,
  DETECTION_FINDING = 2004,
  ACCOUNT_CHANGE = 3001,
  AUTHENTICATION = 3002,
  ENTITY_MANAGEMENT = 3004,
  USER_ACCESS_MANAGEMENT = 3005,
  INVENTORY_INFO = 5001,
  WEB_RESOURCES_ACTIVITY = 6001,
  APPLICATION_LIFECYCLE = 6002,
  API_ACTIVITY = 6003,
  DATASTORE_ACTIVITY = 6005,
  // New control-plane classes in the proposed `ai` category (uid 9).
  // UIDs are provisional pending OCSF issue #1640 ratification.
  AGENT_ACTIVITY = 9001,
  DELEGATION_ACTIVITY = 9002,
}

/**
 * Backward-compatible alias. AITF previously defined a bespoke Category 7 with
 * classes 7001-7010; events now reuse the OCSF classes above per OCSF's
 * object/profile-reuse model. Kept so existing imports keep working.
 */
export const AIClassUID = OCSFClassUID;

/**
 * OCSF `ai_agent.type_id` — normalized agent framework.
 *
 * Mirrors the enum introduced by OCSF PR #1641 (`objects/ai_agent.json`)
 * so AITF telemetry maps cleanly onto the upstream OCSF `ai_agent` object.
 */
export enum AgentTypeID {
  UNKNOWN = 0,
  NATIVE = 1,
  LANGCHAIN = 2,
  AUTOGEN = 3,
  CREWAI = 4,
  OTHER = 99,
}

/** Caption labels for AgentTypeID, per OCSF PR #1641. */
export const AGENT_TYPE_LABELS: Record<number, string> = {
  0: "Unknown",
  1: "Native",
  2: "LangChain",
  3: "AutoGen",
  4: "CrewAI",
  99: "Other",
};

/**
 * AITF framework value -> OCSF ai_agent.type_id. Frameworks without a
 * dedicated OCSF enum member (langgraph, semantic_kernel, custom, ...)
 * normalize to OTHER (99), matching OCSF's open-enum guidance.
 */
const FRAMEWORK_TO_TYPE_ID: Record<string, AgentTypeID> = {
  native: AgentTypeID.NATIVE,
  langchain: AgentTypeID.LANGCHAIN,
  langgraph: AgentTypeID.LANGCHAIN,
  autogen: AgentTypeID.AUTOGEN,
  crewai: AgentTypeID.CREWAI,
};

/** Map an AITF framework string to an OCSF `ai_agent.type_id` value. */
export function normalizeAgentTypeId(framework?: string | null): number {
  if (!framework) {
    return AgentTypeID.UNKNOWN;
  }
  const key = framework.trim().toLowerCase();
  return key in FRAMEWORK_TO_TYPE_ID
    ? FRAMEWORK_TO_TYPE_ID[key]
    : AgentTypeID.OTHER;
}

/**
 * OCSF AI category proposed in OCSF issue #1640. AITF reuses existing OCSF
 * classes for data-plane activity and uses this proposed `ai` category only
 * for the new agent / delegation control-plane classes.
 */
export const OCSF_AI_CATEGORY_UID = OCSFCategoryUID.AI; // proposed "AI Activity" category (OCSF issue #1640)

// --- OCSF Base Object Interfaces ---

/** OCSF event metadata. */
export interface OCSFMetadata {
  version: string;
  product: {
    name: string;
    vendor_name: string;
    version: string;
  };
  uid: string;
  correlation_uid?: string;
  original_time?: string;
  logged_time: string;
}

/** OCSF actor information. */
export interface OCSFActor {
  user?: Record<string, unknown>;
  session?: Record<string, unknown>;
  app_name?: string;
}

/** OCSF device/host information. */
export interface OCSFDevice {
  hostname?: string;
  ip?: string;
  type?: string;
  os?: Record<string, string>;
  cloud?: Record<string, string>;
  container?: Record<string, string>;
}

/** OCSF enrichment data. */
export interface OCSFEnrichment {
  name: string;
  value: string;
  type?: string;
  provider?: string;
}

/** OCSF observable value. */
export interface OCSFObservable {
  name: string;
  type: string;
  value: string;
}

// --- AI-Specific Extension Interfaces ---

/** AI model information. */
export interface AIModelInfo {
  model_id: string;
  name?: string;
  version?: string;
  provider?: string;
  type?: string; // llm, embedding, image, audio, multimodal
  parameters?: Record<string, unknown>;
}

/** AI token usage statistics. */
export interface AITokenUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  cached_tokens: number;
  reasoning_tokens: number;
  estimated_cost_usd?: number;
}

/** AI operation latency metrics. */
export interface AILatencyMetrics {
  total_ms: number;
  time_to_first_token_ms?: number;
  tokens_per_second?: number;
  queue_time_ms?: number;
  inference_time_ms?: number;
}

/** AI operation cost information. */
export interface AICostInfo {
  input_cost_usd: number;
  output_cost_usd: number;
  total_cost_usd: number;
  currency: string;
}

/** Multi-agent team information. */
export interface AITeamInfo {
  team_name: string;
  team_id?: string;
  topology?: string;
  members: string[];
  coordinator?: string;
}

/** Security finding details. */
export interface AISecurityFinding {
  finding_type: string;
  owasp_category?: string;
  risk_level: string;
  risk_score: number;
  confidence: number;
  detection_method: string;
  blocked: boolean;
  details?: string;
  pii_types: string[];
  matched_patterns: string[];
  remediation?: string;
}

/**
 * OCSF `ai_agent` object (OCSF PR #1641).
 *
 * An autonomous AI agent operating under delegated authority. Distinct from
 * the OCSF `agent` object (which models security sensors such as EDR/DLP)
 * and from human principals. Attached to events via the `ai_operation`
 * profile so any activity can be attributed to the agent that performed it.
 */
export interface OCSFAIAgent {
  uid: string; // required: stable logical identifier
  instance_uid?: string; // restart-sensitive running instance id
  name?: string;
  type?: string; // caption of type_id (Native, LangChain, ...)
  type_id: number;
  ai_model?: string; // model backing the agent at event time
  version?: string; // agent code/configuration revision
  charter?: string; // role / operating-boundary reference
}

/**
 * OCSF `delegation` object (OCSF issue #1640).
 *
 * A durable authorization context that persists independently of any single
 * trace or session. `uid`/`parent_uid`/`issuer_uid` provide the OCSF core;
 * the remaining fields preserve AITF's richer delegation telemetry.
 */
export interface OCSFDelegation {
  uid: string; // required: stable delegation identifier
  parent_uid?: string; // parent delegation (lineage)
  issuer_uid?: string; // trusted issuer that minted the delegation
  delegator?: string;
  delegatee?: string;
  type?: string; // on_behalf_of, token_exchange, capability_grant, ...
  scope: string[];
  proof_type?: string; // dpop, mtls_binding, signed_assertion
  ttl_seconds?: number;
}

/** A single node in an OCSF `delegation_lineage` graph (OCSF issue #1640). */
export interface OCSFDelegationNode {
  uid: string;
  parent_uid?: string;
  agent_uid?: string;
  depth?: number;
}

/** OCSF `delegation_lineage` — directed graph for ancestry queries. */
export interface OCSFDelegationLineage {
  nodes: OCSFDelegationNode[];
}

/** Compliance framework mappings. */
export interface ComplianceMetadata {
  nist_ai_rmf?: Record<string, unknown>;
  mitre_atlas?: Record<string, unknown>;
  iso_42001?: Record<string, unknown>;
  eu_ai_act?: Record<string, unknown>;
  soc2?: Record<string, unknown>;
  gdpr?: Record<string, unknown>;
  ccpa?: Record<string, unknown>;
  csa_aicm?: Record<string, unknown>;
}

// --- OCSF Base Event ---

/**
 * Base OCSF event for AITF AI events.
 *
 * Subclasses/factories set `category_uid` and `class_uid` to the OCSF class
 * they reuse (OCSF PR #1641 / issue #1640). AI-specific context is carried on
 * the `ai_operation` profile (`ai_agent`, `ai_model`, `delegation`).
 */
export interface AIBaseEvent {
  activity_id: number;
  category_uid: number;
  class_uid: number;
  type_uid: number;
  time: string;
  severity_id: number;
  status_id: number;
  message: string;
  metadata: OCSFMetadata;
  actor?: OCSFActor;
  device?: OCSFDevice;
  compliance?: ComplianceMetadata;
  observables: OCSFObservable[];
  enrichments: OCSFEnrichment[];

  // OCSF `ai_operation` profile (OCSF PR #1641) + delegation context
  // (OCSF issue #1640). Populated by the crosswalk so every AITF event can
  // be attributed to the AI agent and delegation that produced it.
  ai_agent?: OCSFAIAgent;
  ai_model?: string;
  delegation?: OCSFDelegation;
  delegation_lineage?: OCSFDelegationLineage;
}

// --- Factory Functions ---

/** Create default OCSF metadata. */
export function createMetadata(
  correlationUid?: string
): OCSFMetadata {
  return {
    version: "1.1.0",
    product: {
      name: "AITF",
      vendor_name: "AITF",
      version: "1.0.0",
    },
    uid: randomUUID(),
    correlation_uid: correlationUid,
    logged_time: new Date().toISOString(),
  };
}

/** Create default token usage. */
export function createTokenUsage(
  options: Partial<AITokenUsage> = {}
): AITokenUsage {
  const usage: AITokenUsage = {
    input_tokens: options.input_tokens ?? 0,
    output_tokens: options.output_tokens ?? 0,
    total_tokens: options.total_tokens ?? 0,
    cached_tokens: options.cached_tokens ?? 0,
    reasoning_tokens: options.reasoning_tokens ?? 0,
    estimated_cost_usd: options.estimated_cost_usd,
  };
  if (usage.total_tokens === 0) {
    usage.total_tokens = usage.input_tokens + usage.output_tokens;
  }
  return usage;
}

/** Create a base OCSF AI event. */
export function createBaseEvent(
  classUid: number,
  options: Partial<AIBaseEvent> = {}
): AIBaseEvent {
  const activityId = options.activity_id ?? OCSFActivity.OTHER;
  return {
    activity_id: activityId,
    // Default to APPLICATION (6); factories override with the reused OCSF
    // category for the class they emit (OCSF PR #1641 / issue #1640).
    category_uid: options.category_uid ?? OCSFCategoryUID.APPLICATION,
    class_uid: classUid,
    type_uid: options.type_uid ?? classUid * 100 + activityId,
    time: options.time ?? new Date().toISOString(),
    severity_id: options.severity_id ?? OCSFSeverity.INFORMATIONAL,
    status_id: options.status_id ?? OCSFStatus.SUCCESS,
    message: options.message ?? "",
    metadata: options.metadata ?? createMetadata(),
    actor: options.actor,
    device: options.device,
    compliance: options.compliance,
    observables: options.observables ?? [],
    enrichments: options.enrichments ?? [],
  };
}

/**
 * Remove undefined/null properties from an object (for JSON export).
 */
export function stripNulls<T extends Record<string, unknown>>(
  obj: T
): Partial<T> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value !== undefined && value !== null) {
      if (
        typeof value === "object" &&
        !Array.isArray(value) &&
        value !== null
      ) {
        const stripped = stripNulls(value as Record<string, unknown>);
        if (Object.keys(stripped).length > 0) {
          result[key] = stripped;
        }
      } else {
        result[key] = value;
      }
    }
  }
  return result as Partial<T>;
}
